### ISSUE TITLE
Inconsistent closure of socket ? EthStratumClient::disconnect() 

### ISSUE THREAD

**Issue by AndreaLanfranchi** [ethereum-mining/ethminer#1025]
I've noticed that on disconnection of ethminer keeps sending some data to remote endpoint. This is endorsed by the fact that some nonces are sent to remote endpoint even after socket disconnection.

I think (please be patient as my C++ is very rusted) proper disconnection should include a shutdown before socket closure for both secure and non-secure sockets.

As this sub is called on every connection error, some error checking against ec should be performed: at present any error is simply ignored.

```
void EthStratumClient::disconnect()
{
	m_worktimer.cancel();
	m_responsetimer.cancel();
	m_response_pending = false;
	m_linkdown = true;

	try {
		
		boost::system::error_code ec;
		
		if (m_connection.SecLevel() != SecureLevel::NONE) {
			m_securesocket->shutdown(boost::asio::ip::tcp::socket::shutdown_both, ec);
			m_securesocket->close();
		else {
			m_socket->shutdown(boost::asio::ip::tcp::socket::shutdown_both, ec);
			m_socket->close();
		}
		
		m_io_service.stop();
	}
	catch (std::exception const& _e) {
		cwarn << "Error while disconnecting:" << _e.what();
	}

        if (m_connection.SecLevel() != SecureLevel::NONE) {
                if (m_securesocket)
                        m_securesocket = nullptr;
        }
        else {
                if (m_socket)
                        m_socket = nullptr;
        }

	m_authorized = false;
	m_connected.store(false, std::memory_order_relaxed);

	if (m_onDisconnected) {
		m_onDisconnected();
	}
}
```

**Comment by AndreaLanfranchi** [comment]
@chfast

**Comment by chfast** [comment]
I don't know much about this piece of code. Thanks for reporting the issue.

**Comment by chfast** [comment]
I think the lifetime of the connection is handled incorrectly.

**Comment by invidtiv** [comment]
I don t know if this is related , but my guess is yes, after trying to exit etherminer or trying to restart. on dev4.
It does not and continues to submit nouce to the pool after stating 
Shutting down 
Disconnected from eu1.ethermine.org
Shutting down miners

**Comment by AndreaLanfranchi** [comment]
All addressed in #1027 which is merged in 0.14
